### ISSUE TITLE
RDom bounds are garbled for IIR filter example if filter order is 10 or more

### ISSUE THREAD

**Issue by gchauras** [halide/Halide#601]
Reproducing test case: https://github.com/gchauras/halide-rdom-coden-bug

IIR filter of n order on 1D data.
```
f(x) = input(x)
f(x) = f(x) + a1*f(x-1) + a2*f(x-2) + ... + an*f(x-n)
```
RDom used to scan the data from left to right - RDom bounds are 0 to image width. 

Everything is fine for order `n<=9`. For `n>9`, the rdom bounds in codegen start from 10 instead of 0. The codgen is dumped out as `stmt_order_<filter order>.html`

Running:
```
./test_app [filter order]
```

**Comment by gchauras** [comment]
@abadams about your comment regarding manually peeling the update defs to remove the selects and max/min: for GPU schedules new update defs will add more syncthreads. I guess this will be more expensive than letting the computing the max/min etc? For CPU schedules it makes sense of course. I am just contemplating if I should go down this road or not.

**Comment by abadams** [comment]
If the serial for loop that you're peeling is inside of the loop over
threads it shouldn't introduce any synchronization. If it's outside, then
it will. We inject syncthreads when there's a sequence point outside of a
notional loop over gpu threads.

**Comment by abadams** [comment]
Hopefully fixed by #611 ?

**Comment by kree-colemcalughlin** [comment]
I'm back from the holidays, so Happy New Year! Thanks for pointing this issue out I have pushed a PR #611 that should fix the problem. A few comments about this:

1) The branching_limit of 10 is a somewhat arbitrary, hard coded value currently that is required to avoid combinatorial explosion of code in certain cases. The case you are presenting provides some incentive for allowing users to adjust this limit for certain programs. However, we don't have a system for introducing some kind of compile-time parameters to Halide that would let us do this.

2) The solution I've pushed up allows halide to generate equivalent stmts for all orders in your example, however it starts producing somewhat unexpected results for orders > 10. Basically, what happens is once it reaches the limit of 10 it bails out and produces a single branch. Then it starts peeling of individual stmts again starting with the 11th store. So for order 12, for example, you will se a loop of extent 10 followed by a single store, followed by the remaining select-free loop of extent 23. I'm not sure there is an easy solution to this problem as it has to do with how the select and add nodes are nested in the IR tree, indeed if you reverse the order of your loop over k you will see different results. Of course, all of this would be made irrelevant if we invest in a solution to 1).

**Comment by zvookin** [comment]
If the limit can vary on a Func by Func basis, the solution would be to add
a scheduling directive.  If not, we could consider adding a field to the
Target structure. Both are fairly easy to do.

-Z-

**Comment by kree-colemcalughlin** [comment]
Yeah, I don't think it is a parameter that is worth a scheduling directive or target field. What I had in mind is a system where you could provide optional compilation parameters to the func.compile_to_*() methods, that allow advanced users to fine tune the compilation of Halide code. These could be useful for adjusting parameters on individual passes, or even turning on/off optional optimization passes. In this case the so-called branching limit feels a lot like the limits on automatic loop unrolling you find in many compilers. It is a parameter that would probably almost never be touched, but could be useful to expose for some edge cases. Doesn't seem like a huge priority though.

**Comment by abadams** [comment]
I think you just described the Target struct :)

**Comment by kree-colemcalughlin** [comment]
Point taken, although the target struct only seems to handle bool valued options right now. I guess the main difference I had in mind is a more dynamic structure where the writer of an optimization pass would register a parameter with an options structure, so that one would not have to be constantly changing the Target struct and updating the parsing routines to reflect new parameters.

**Comment by abadams** [comment]
We've been on the verge of needing a std::map<string, Variant> in Target
for a while. Perhaps this is what tips us over.

**Comment by gchauras** [comment]
Issue fixed by #611. 
I am leaving this open because this looks a conversation in progress. Feel free to close this an move the discussion elsewhere. Thanks.
